File: Leetcode/04_Strings/026_0028_find-the-index-of-the-first-occurrence-in-a-string.go
// https://leetcode.cn/problems/find-the-index-of-the-first-occurrence-in-a-string/description/
// e

package main

// KMP 算法
func strStr(haystack string, needle string) int {
	n := len(needle)
	if n == 0 {
		return 0
	}
	next := getNext(needle)
	hi, ni := 0, 0
	for hi < len(haystack) {
		for hi < len(haystack) && ni < n && haystack[hi] == needle[ni] {
			hi++
			ni++
		}
		if ni == n {
			return hi - n
		}
		if ni > 0 {
			ni = next[ni-1]
		} else {
			hi++
		}
	}
	return -1
}

func getNext(str string) (next []int) {
	next = make([]int, len(str))
	for i, j := 0, 1; j < len(next); j++ {
		for str[i] != str[j] && i > 0 {
			i = next[i-1]
		}
		if str[i] == str[j] {
			i++
		}
		next[j] = i
	}
	return
}

File: Leetcode/04_Strings/026_0028_find-the-index-of-the-first-occurrence-in-a-string.py
class Solution(object):
    def strStr(self, haystack, needle):
        if needle == '':
            return 0
        nxt = self.getNext(needle)
        i = j = 0
        while j < len(haystack):
            while j < len(haystack) and i < len(needle) and needle[i] == haystack[j]:
                i += 1
                j += 1
            if i == len(needle):
                return j - i
            if i > 0:
                i = nxt[i-1]
            else:
                j += 1
        return -1
    
    def getNext(self, pstr):
        i = 0
        nxt = [0] * len(pstr)
        for j in range(1, len(pstr)):
            while pstr[i] != pstr[j] and i > 0:
                i = nxt[i-1]
            if pstr[i] == pstr[j]:
                i += 1
            nxt[j] = i
        return nxt

s = Solution()
print(s.strStr('abc','c'))